File: src/resources/clients/java_client/src/visitclient/VisItProxy.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package visitclient;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.ConnectException;
import java.net.Socket;

/**
 *
 * @author hari
 */
public class VisItProxy
{

    public static final int BUFSIZE = 4096;
    
    class Header
    {
        public String password;
        public String canRender;
    }
        
    class VisItThread implements Runnable
    {
        private InputStreamReader inputConnection;
        private Gson gson;
        
        VisItThread(InputStreamReader i)
        {
            gson = new Gson();
            inputConnection = i;
        }
        
        private int count(String str, String findStr)
        {
            int lastIndex = 0;
            int count =0;

            while(lastIndex != -1)
            {

                lastIndex = str.indexOf(findStr,lastIndex);

                if( lastIndex != -1){
                    count ++;
                    lastIndex+=findStr.length();
                }
            }
            
            return count;
        }
        
        @Override
        public void run()
        {
            StringBuilder partial_entry = new StringBuilder(""); //stitches together one map_node entry
            StringBuilder input_buffer = new StringBuilder(""); //holds input data buffer

            char [] data = new char [ VisItProxy.BUFSIZE ];
            
            while(true)
            {
                int len = 0;
                try {
                    len = inputConnection.read(data);
                }
                catch(Exception e) 
                {
                    System.out.println("Exception: Quitting ...");
                    break;
                }
                
                if(len == 0)
                {
                    System.out.println("Quitting ...");
                    break;
                }
                
                input_buffer.append(data,0,len);
                
                // for now JSON parser has to start with object..
                int mnsi = input_buffer.indexOf("{");
                int mnei = input_buffer.indexOf("}");

                while (mnsi >= 0 || mnei >= 0)
                {
                    if (mnsi < 0 && mnei >= 0)
                    {
                        mnei += "}".length();
                        partial_entry.append(input_buffer.subSequence(0,mnei));
                        input_buffer.delete(0, mnei);
                    }
                    else if (mnsi >= 0 && mnei < 0)
                    {
                        mnsi += "{".length();
                        partial_entry.append(input_buffer.subSequence(0,mnsi));
                        input_buffer.delete(0, mnsi);
                    }
                    else
                    {
                        if( mnsi < mnei )
                        {
                            mnsi += "{".length();
                            partial_entry.append(input_buffer.subSequence(0, mnsi));
                            input_buffer.delete(0,mnsi);
                        }
                        else
                        {
                            mnei += "}".length();
                            partial_entry.append(input_buffer.subSequence(0,mnei));
                            input_buffer.delete(0,mnei);
                        }
                    }
                
                    String tmp = partial_entry.toString().trim();

                    if (count(tmp,"{") > 0 && count(tmp,"{") == count(tmp,"}"))
                    {
                        try {
                            partial_entry.setLength(0);
                            //tmp=tmp.replace("\n","");
                            //tmp=tmp.replace("\\\"","");
                            JsonElement el = gson.fromJson(tmp,JsonElement.class);
                            JsonObject jo = el.getAsJsonObject();

                            ///update state..
                            VisItProxy.this.state.update(jo);
                        }
                        catch(Exception e)
                        {
                            System.out.println("failed input " + tmp);
                        }
                    }

                    //tmp = input_buffer.strip()

                    //print partial_entry
                    mnsi = input_buffer.indexOf("{"); //don't include current node
                    mnei = input_buffer.indexOf("}");
                }
            }
        }
    }
    
    private boolean handshake(String host, int port, String password)
    {
        try {
            Gson gson = new Gson();
            
            Socket socket = new Socket(host,port);
            OutputStreamWriter writer = new OutputStreamWriter(socket.getOutputStream());
            InputStreamReader reader = new InputStreamReader(socket.getInputStream());
           
            Header header = new Header();
            header.password = password;
            header.canRender = "data";
            
            String headerstr = gson.toJson(header);
//            System.out.println("Writing message: " + headerstr);
            
            writer.write(headerstr);
            writer.flush();
            
//            System.out.println("Reading...");
            char [] cbuf = new char [ 1024 ];
            int len = reader.read(cbuf);
            String message = new String(cbuf,0,len);
           
            JsonElement e = gson.fromJson(message,JsonElement.class);
            JsonObject jo = e.getAsJsonObject();
 
            visit_host = jo.get("host").getAsString();
            visit_port = jo.get("port").getAsString();
            visit_security_key = jo.get("securityKey").getAsString();
            
            //System.out.println(visit_host + " " + visit_port + " " + visit_security_key);
            socket.close();
            
            return true;
        }
        catch(ConnectException e)
        {
            System.out.println("Exception: Could not connect to a running VisIt client");
        }
        catch(Exception e)
        {
            System.out.println("Exception");
        }
        return false;
    }
    
    public boolean connect(String host, int port, String password)
    {
        if(!handshake(host, port, password))
            return false;
        
        try{
            
            inputSocket = new Socket(visit_host,Integer.valueOf(visit_port));
            inputConnection = new InputStreamReader(inputSocket.getInputStream());
            
            outputSocket = new Socket(visit_host,Integer.valueOf(visit_port));
            outputConnection = new OutputStreamWriter(outputSocket.getOutputStream());
            
            
            // Handle initial connection 
            char [] cbuf = new char [1024];
            
            //read 100 bytes
            InputStreamReader isr = (new InputStreamReader(outputSocket.getInputStream()));
            isr.read(cbuf);
            
            //extract socket key
            //visit_socket_key = new String(cbuf,5+1+10+21, 21);
            
            cbuf[0] = ASCIIFORMAT;
            
            for(int i = 0; i < visit_security_key.length(); ++i)
                cbuf[6+10+i] = visit_security_key.charAt(i);
            
            OutputStreamWriter osw = new OutputStreamWriter(inputSocket.getOutputStream());
            osw.write(cbuf);
            osw.flush();
        
            //System.out.println("wrote: " + new String(cbuf));
            // End - Handle initial connection
            
            state.setConnection(outputConnection);
            
            thread = new Thread(new VisItThread(inputConnection));
            
            thread.setDaemon(true);
            thread.start();
            
            //Hackish sync state..
            while(state.states.size() < 135)
            {
                try{Thread.sleep(100); }catch(Exception e) {}
            }

            System.out.println("Viewer State synched..");

        }
        catch(Exception e)
        {
            System.out.println("Exception connecting..");
        }
        return true;
    }
    
    public VisItProxy()
    {
        state = new ViewerState();
        methods = new ViewerMethods(state);
    }
    
    public ViewerState GetViewerState()
    {
        return state;
    }
    
    public ViewerMethods GetViewerMethods()
    {
        return methods;
    }
    
    private String visit_host, visit_port;
    //private String visit_socket_key;
    private String visit_security_key;
    private InputStreamReader inputConnection;
    private OutputStreamWriter outputConnection;
    private Socket inputSocket, outputSocket;
    private Thread thread;
   
    private static final byte ASCIIFORMAT = 0;
    //private static final byte BINARYFORMAT = 1;

    private ViewerState state;
    private ViewerMethods methods;
}

File: src/resources/clients/python_client/visitproxy.py
import time
import socket
if (sys.version_info > (3, 0)):
    import _thread
else:
    import thread as _thread
import threading
import visitstate
import atexit
import inspect
import json
import sys

from viewerstate import *
from viewermethods import *

# Event records reciept of last VisIt API call
LastEvent = False

class ViewerProxy:
    ASCIIFORMAT=0
    BINARYFORMAT=1
    BUFSIZE = 4096

    def __init__(self):

        self.state = ViewerState()
        self.methods = ViewerMethods(self.state)

        #internal data
        self.visit_host = ""
        self.visit_port = -1
        self.visit_security_key = ""
        self.visit_socket_key = ""
        self.inputConnection = None
        self.outputConnection = None
        self.headerData = bytearray('')
        self.globalList = None

    #create syncState before anything else
    #SyncCount = 0
    #def handleCallback(self,subject):
    #    if subject.typename == "SyncAttributes":
    #        #ignore first since that is initial connection
    #        if ViewerProxy.SyncCount >= 1 and subject.data["syncTag"] == 10000:
    #            print "releasing lock",subject.data["syncTag"]
    #        ViewerProxy.SyncCount += 1

    #def synchronize(self,value):
    #    #send VisIt sync signal..
    #    self.state.data(2)["syncTag"] = int(value)
    #    self.state.notify(2,"SyncAttributes")
    #    self.state.data(2)["syncTag"] = -1
    #    self.event.wait()

    def handshake(self,host,port,password,visType):
        # connect to visit given host, port, and password
        # VisIt must be listener mode for this to work
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        # connect to VisIt
        s.connect((host,port))

        # VisIt expects password to continue
        header = {}
        header["password"] = password
        header["canRender"] = visType
        s.send(json.dumps(header))

        data = s.recv(ViewerProxy.BUFSIZE) # read handshake message

        s.close() # close connection after message

        if not data:
            print("Connection to VisIt failed")
            return False

        message = json.loads(data)
        self.visit_host = str(message["host"])
        self.visit_port = int(message["port"])
        self.visit_security_key = str(message["securityKey"])
        return True

    # Header information contains:
    # 5 bytes for format information
    # 1 byte for failcode,
    # 10 bytes for version number
    # 21 bytes for security key
    # 21 bytes for socket key
    
    def readHeader(self,conn):

        # the header should be exactly 100 bytes
        self.headerData = conn.recv(ViewerProxy.BUFSIZE)

        #computed from header info (not really needed, but kept for debugging)
        self.visit_socket_key = self.headerData[5+1+10+21:5+1+10+21+21]

    def writeHeader(self,conn):
        # modify header data for response..
        response = bytearray(self.headerData)
        
        # set ascii format as response type
        response[0] = ViewerProxy.ASCIIFORMAT
        
        # write security key in response
        for i in range(len(self.visit_security_key)):
            response[5+1+10+i] = self.visit_security_key[i] #21st byte is NULL?

        # send 100 byte response
        conn.send(response)

    #TODO: I need a way of telling that the last api has come in
    #before adding the all of VisIt's functions to the global modules

    count = 0
    def updateEvent(self,subject):
        global LastEvent

        if subject.typename == "ClientInformationList":
            self.count = self.count + 1
            if self.count == 3:
                LastEvent = True

    def connect(self,host,port,password,visType):
        global LastEvent

        # attempt handshake
        if not self.handshake(host,port,password,visType):
            print("HandShake failed: unable to connect")
            return False

        #self.visit_host = "localhost";
        # create connections from handshake information
        self.inputConnection = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
        self.inputConnection.connect((self.visit_host,self.visit_port))

        self.outputConnection = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
        self.outputConnection.connect((self.visit_host,self.visit_port))

        # read header
        self.readHeader(self.outputConnection)

        # write header
        self.writeHeader(self.inputConnection)

        #give state the connection
        self.state.setConnection(self.outputConnection)

        self.state.addListener(self.updateEvent)

        #run eventloop
        _thread.start_new_thread(self.eventLoop,())

        #wait until VisIt API has loaded
        #TODO: replace this logic with semaphore wait..
        print("Loading VisIt API...")
        while(not LastEvent): time.sleep(.1)

        self.state.removeListener(self.updateEvent)

        self.sync()
        print("The viewer proxy has loaded...")

    def close(self):
        self.inputConnection.close()
        self.outputConnection.close()

    def notify(self,index,tag=""):
        self.state.notify(index,tag)

    def printXml(self,index,tag=""):
        self.state.printXml(index,tag)

    def sync(self):
            gmap = sys.modules["__main__"].__dict__
            if type(gmap) is not dict : return

            for i in range(len(self.state.states)):
                typename = self.state.states[i].typename
                if typename not in gmap:
                    gmap[typename] = self.state.states[i].sync() 
            #add viewer method functions
            self.methods.sync()

    def eventLoop(self):
        partial_entry = "" #stitches together one map_node entry
        input_buffer = "" #holds input data buffer

        while True:
            data = self.inputConnection.recv(ViewerProxy.BUFSIZE)

            if not data:
                print("Quitting ...")
                break;
        
            input_buffer += data
            
            # for now JSON parser has to start with object..
            mnsi = input_buffer.find("{");
            mnei = input_buffer.find("}");

            while mnsi >= 0 or mnei >= 0:
                if mnsi < 0 and mnei >= 0:
                    mnei += len("}")
                    partial_entry += input_buffer[0:mnei]
                    input_buffer = input_buffer[mnei:]
                elif mnsi >= 0 and mnei < 0:
                    mnsi += len("{")
                    partial_entry += input_buffer[0:mnsi]
                    input_buffer = input_buffer[mnsi:]
                else:
                    if mnsi < mnei:
                        mnsi += len("{")
                        partial_entry += input_buffer[0:mnsi]
                        input_buffer = input_buffer[mnsi:]
                    else:
                        mnei += len("}")
                        partial_entry += input_buffer[0:mnei]
                        input_buffer = input_buffer[mnei:]
                
                tmp = partial_entry.strip()

                if tmp.count("{") > 0 and tmp.count("{") == tmp.count("}"):
                    try:
                        partial_entry = ""
                        tmp=tmp.replace("\n","")
                        tmp=tmp.replace("\\\"","")
                        res = json.loads(tmp)
                        self.state.update(res)
                    except:
                        print("failed input", tmp)

                tmp = input_buffer.strip()

                #print partial_entry
                mnsi = input_buffer.find("{") #don't include current node
                mnei = input_buffer.find("}")
    

